precompiles/ics20/approve_common.go

```diff
@@ -0,0 +1,267 @@
+// Copyright Tharsis Labs Ltd.(Evmos)
+// SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/evmos/blob/main/LICENSE)
+
+package ics20
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/evmos/evmos/v14/precompiles/authorization"
+
+	errorsmod "cosmossdk.io/errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/authz"
+	authzkeeper "github.com/cosmos/cosmos-sdk/x/authz/keeper"
+	transfertypes "github.com/cosmos/ibc-go/v7/modules/apps/transfer/types"
+	channelkeeper "github.com/cosmos/ibc-go/v7/modules/core/04-channel/keeper"
+	channeltypes "github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+	cmn "github.com/evmos/evmos/v14/precompiles/common"
+)
+
+// TransferMsgURL is the ICS20 transfer message URL string.
+var TransferMsgURL = sdk.MsgTypeURL(&transfertypes.MsgTransfer{})
+
+// Approve implements the ICS20 Authorization approve transactions.
+func Approve(
+	ctx sdk.Context,
+	authzKeeper authzkeeper.Keeper,
+	channelKeeper channelkeeper.Keeper,
+	precompileAddr, grantee, origin common.Address,
+	approvalExpiration time.Duration,
+	transferAuthz *transfertypes.TransferAuthorization,
+	event abi.Event,
+	stateDB vm.StateDB,
+) error {
+	// If one of the allocations contains a non-existing channel, throw and error
+	for _, allocation := range transferAuthz.Allocations {
+		found := channelKeeper.HasChannel(ctx, allocation.SourcePort, allocation.SourceChannel)
+		if !found {
+			return errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "port ID (%s) channel ID (%s)", allocation.SourcePort, allocation.SourceChannel)
+		}
+	}
+
+	// Only the origin can approve a transfer to the grantee address
+	expiration := ctx.BlockTime().Add(approvalExpiration).UTC()
+	if err := authzKeeper.SaveGrant(ctx, grantee.Bytes(), origin.Bytes(), transferAuthz, &expiration); err != nil {
```

Comment:
should we also call this `granter` instead of `origin`? We had some confusion around this in the past